src/api.ts

```diff
@@ -0,0 +1,395 @@
+}
+
+/**
+ * An AWS API Gateway integration lets you integrate an API with an AWS service
+ * supported by Functionless. The request is transformed via VTL and sent to the
+ * service via API call, and the response is transformed via VTL and returned in
+ * the response.
+ *
+ * TODO: we need to support multiple responses
+ *
+ * Only `application/json` is supported.
+ *
+ * TODO: provide example usage after api is stabilized
+ *
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-api-integration-types.html
+ */
+export class AwsApiIntegration<
+  Props extends AwsApiIntegrationProps<any, any, any, any>
+> implements BaseApiIntegration
+{
+  /**
+   * This static property identifies this class as an AwsApiIntegration to the Functionless plugin.
+   */
+  public static readonly FunctionlessType = "AwsApiIntegration";
+
+  private readonly request: FunctionDecl;
+  private readonly response: FunctionDecl;
+  private readonly integration: Props["integration"];
+
+  constructor(props: Props) {
+    this.request = validateFunctionDecl(props.request);
+    this.response = validateFunctionDecl(props.response);
+    this.integration = props.integration;
+  }
+
+  addMethod(httpMethod: HttpMethod, resource: aws_apigateway.Resource): void {
+    const requestTemplate = toVTL(this.request, "request");
+    const responseTemplate = toVTL(this.response, "response");
+
+    const apiGWIntegration = this.integration.apiGWVtl.integration(
+      requestTemplate,
+      responseTemplate
+    );
+
+    // TODO: support requestParameters, authorizers, models and validators
+    resource.addMethod(httpMethod, apiGWIntegration, {
+      methodResponses: [
+        {
+          statusCode: "200",
+        },
+      ],
+    });
+  }
+}
+
+/**
+ * Simple VTL interpreter for a FunctionDecl. The interpreter in vtl.ts
+ * is based on the AppSync VTL engine which has much more flexibility than
+ * the API Gateway VTL engine. In particular it has a toJson utility function
+ * which means the VTL can just create an object in memory and then toJson it
+ * the end. Here we need to manually output the JSON which is how VTL is
+ * typically meant to be used.
+ *
+ * For now, only Literals and references to the template input are supported.
+ * It is definitely possible to support more, but we will start with just
+ * small core and support more syntax carefully over time.
+ *
+ * @param node Function to interpret.
+ * @param template Whether we are creating a request or response mapping template.
+ */
+function toVTL(node: FunctionDecl, template: "request" | "response") {
+  const statements = node.body.statements.map((stmt) => inner(stmt)).join("\n");
+
+  if (template === "request") {
+    return `#set($inputRoot = $input.path('$'))${statements}`;
+  } else {
+    return statements;
+  }
+
+  function inner(node: FunctionlessNode): string {
+    switch (node.kind) {
+      case "ArrayLiteralExpr":
```

Comment:
We are moving the entire code base to if-else with the `is*Expr` assertion functions, eg `isArrayLiteralExpr`.